clar2wasm/src/error_mapping.rs

```diff
@@ -0,0 +1,157 @@
+            5 => ErrorMap::UnwrapFailure,
+            6 => ErrorMap::Panic,
+            7 => ErrorMap::ShortReturnAssertionFailure,
+            8 => ErrorMap::ArithmeticPowError,
+            _ => ErrorMap::NotMapped,
+        }
+    }
+}
+
+pub(crate) fn resolve_error(
+    e: wasmtime::Error,
+    instance: Instance,
+    mut store: impl AsContextMut,
+) -> Error {
+    if let Some(vm_error) = e.root_cause().downcast_ref::<Error>() {
+        // SAFETY:
+        //
+        // This unsafe operation returns the value of a location pointed by `*mut T`.
+        //
+        // The purpose of this code is to take the ownership of the `vm_error` value
+        // since clarity::vm::errors::Error is not a Clonable type.
+        //
+        // Converting a `&T` (vm_error) to a `*mut T` doesn't cause any issues here
+        // because the reference is not borrowed elsewhere.
+        //
+        // The replaced `T` value is deallocated after the operation. Therefore, the chosen `T`
+        // is a dummy value, solely to satisfy the signature of the replace function
+        // and not cause harm when it is deallocated.
+        //
+        // Specifically, Error::Wasm(WasmError::ModuleNotFound) was selected as the placeholder value.
+        return unsafe {
+            core::ptr::replace(
+                (vm_error as *const Error) as *mut Error,
+                Error::Wasm(WasmError::ModuleNotFound),
+            )
+        };
+    }
+
+    if let Some(vm_error) = e.root_cause().downcast_ref::<CheckErrors>() {
+        // SAFETY:
+        //
+        // This unsafe operation returns the value of a location pointed by `*mut T`.
+        //
+        // The purpose of this code is to take the ownership of the `vm_error` value
+        // since clarity::vm::errors::Error is not a Clonable type.
+        //
+        // Converting a `&T` (vm_error) to a `*mut T` doesn't cause any issues here
+        // because the reference is not borrowed elsewhere.
+        //
+        // The replaced `T` value is deallocated after the operation. Therefore, the chosen `T`
+        // is a dummy value, solely to satisfy the signature of the replace function
+        // and not cause harm when it is deallocated.
+        //
+        // Specifically, CheckErrors::ExpectedName was selected as the placeholder value.
+        return unsafe {
+            let err = core::ptr::replace(
+                (vm_error as *const CheckErrors) as *mut CheckErrors,
+                CheckErrors::ExpectedName,
+            );
+
+            <CheckErrors as std::convert::Into<Error>>::into(err)
+        };
+    }
+
+    // Check if the error is caused by
+    // an unreachable Wasm trap.
+    //
+    // In this case, runtime errors are handled
+    // by being mapped to the corresponding ClarityWasm Errors.
+    if let Some(Trap::UnreachableCodeReached) = e.root_cause().downcast_ref::<Trap>() {
+        return from_runtime_error_code(instance, &mut store);
+    }
```

Comment:
I still have a problem with this: if we have a real `unreachable`, it will be treated like it was a runtime error, which will be `NotMapped`. This kind of error should happen only during development, but it will slow us down if we cannot receive the actual error.

Reply:
I'm not sure I understand your point completely. 

By real unreachable do you mean an `unreachable!()` that causes a panic? The condition is only considering a `wasmtime::Trap:: UnreachableCodeReached`, isn't it?

Reply:
Changed this conditional to also check for the global wasm runtime-error. If it is `-1` it means that it was thrown by a `real unreachable`.